### DISCUSSION TITLE
posts/how-to-report-progress/

### DISCUSSION THREAD

**Discussion by giscus** [BYJRK/byjrk.github.io#10] [Announcements]
# posts/how-to-report-progress/

在执行异步任务时，有时候我们会希望有办法汇报进度。比如在一个 WPF 程序中，我们在界面上放了一个进度条，从而展示当前任务的进度。那么该如何汇报异步

https://blog.coldwind.top/posts/how-to-report-progress/

**Comment by lxymahatma** [comment]
想问个和 Progress 相关性不是很大的问题
我理解的话就是这里因为要直接通过 `progressBar.Value` 这样的方式来直接访问进度条，所以 `DoJobAsync` 这个方法就写到了 MainWindow 里而不是使用 ViewModel，然后因为 Button 点击事件需要调用这个 `DoJobAsync` 方法，所以 Button 的点击事件也只能是UI的点击事件，最终导致了使用 `async void`
如果使用 MVVM 的模式，因为可以通过 command binding 来绑定异步方法，某种程度上应该是比 `async void` 要安全一点。那么在写类似程序的时候如果要一个进度条，是先考虑 MVVM 的模式还是直接通过控件名字访问修改值这样的方法更好 （假设还在设计阶段，因为如果其他地方已经用了 MVVM 就不用考虑这个了）
也就是某种程度上说，是不是 MVVM 设计模式更加先进安全一点

**Reply by fangqingyao** [reply]
你的想法和我一样，MVVM 的模式progressBar.Value可以Bing ViewModel里面的一个属性，可能就可以更好的避免跨线程访问控件。

**Reply by BYJRK** [reply]
不能这么说。MVVM 本身并没有任何提供线程安全的意图。实现线程安全的是 WPF 本身的绑定功能。而且即便你使用了绑定，在 VM 中你仍然需要考虑线程安全等问题

**Comment by lxymahatma** [comment]
> 不能这么说。MVVM 本身并没有任何提供线程安全的意图。实现线程安全的是 WPF 本身的绑定功能。而且即便你使用了绑定，在 VM 中你仍然需要考虑线程安全等问题

所以是说，`async void` 和 `async Task` 方法里该有的 try catch 都是一个不能少，但是 `async Task` 好处就是如果你忘了 try catch 它的报错会更加详细（？

**Reply by BYJRK** [reply]
不能这样说。他们俩的本质区别是 `async void` 根本就没有返回一个 `Task` 对象去存储异步任务的各种状态，包括是否完成、是否出错等。一旦出错，也不知道该怎么办，只能直接抛出。

**Comment by mafei235** [comment]
upup求B站视频《WPF 中如何将一组 RadioButton 绑定到 ViewModel 的属性》https://www.bilibili.com/video/BV1E14y137qd/?spm_id_from=333.880.my_history.page.click&vd_source=6be89e62ac6158bdff73d716e4c3cb14    的源代码。视频下提供的GitHub代码地址失效了。感谢。

**Reply by BYJRK** [reply]
我回头看一下。这个年代有点久远了

**Reply by mafei235** [reply]
我也在B站视频下留言了。

**Comment by MoonHidesNight** [comment]
BackgroundWorker这个类直接可以实现这种需求吧，博主
